### DISCUSSION TITLE
Virtual Characters and Sensors

### DISCUSSION THREAD

**Discussion by RoboMage** [jrouwe/JoltPhysics#856] [Q&A] [answered]
Is there a way for Virtual Characters to detect collisions with sensors, without using a second kinematic body? I noticed in CharacterContactListener when I have a contact and cross check the BodyID with the bodies in my scene, that the only ones that never appear are bodies I have called SetIsSensor(true) on.  I confirmed sensors are working with other dynamic bodies in the scene. 

I followed the advice from some of the other discussions and added another body that I sync to my VirtualCharacter, but it seems weird to me to add all this code just to make virtual characters activate triggers.

Could someone explain the proper usage of sensors in Jolt?
Thanks :)

**Comment by TIHan** [comment, marked as answer]
I think adding the non-virtual character is the only way to detect collisions with sensors.

To make this scenario less confusing, maybe we should add a new test/example that demonstrates how to use the combined VirtualCharacter + Character properly.

**Reply by RoboMage** [reply]
I feel like it would be really nice if virtual characters were sufficient; but yeah I definitely agree a new example explaining sensors would be awesome.

**Reply by jrouwe** [reply]
I never needed CharacterVirtual to interact with sensors, but it is quite simple to implement. I've created #861.

Making a demo that combines a virtual character + character has been on my (rather large) to do list for a while.

**Reply by TIHan** [reply]
I'm happy to take a look at adding the example this weekend. Is there anything specific other than just making the character/virtual character together?

**Reply by jrouwe** [reply]
Well, I haven't really decided what form this should take. One possibility is that CharacterVirtual will just get the possibility to have a Character attached to it and it will move it along whenever it moves. Another is that there will be a utility class that syncs both. A third possibility is that it will only be a demo and that it will be up to people to copy the implementation.

**Reply by jankrassnigg** [reply]
You recently pushed a change, which allows us to see contacts with triggers, right? So one could handle those directly to determine whether a character is in a trigger zone, or is it not enough for that?

And on a related note, what is the benefit to having a rigid-body character simultaneously with a virtual one? I currently just use a second kinematic body, which takes care of pushing things away, I don't see how a simulated body would improve this?

**Reply by jrouwe** [reply]
> So one could handle those directly to determine whether a character is in a trigger zone, or is it not enough for that?

Yes that's the point of the change.

> And on a related note, what is the benefit to having a rigid-body character simultaneously with a virtual one? I currently just use a second kinematic body, which takes care of pushing things away, I don't see how a simulated body would improve this?

If I remember correctly, what we use is a dynamic body with a slightly smaller capsule with GravityFactor = 0 in an ObjectLayer that collides only with dynamic objects and we move it using SetLinearVelocity (which we call soft keyframing). The advantage of this is that it can push stuff but not with infinite force. It uses the Character class because at the time it was the only thing that implemented an inertia that didn't allow for rotation, but nowadays that's just settable by EAllowedDOFs. So I guess you're right, I could just add support for a regular rigid body to the CharacterVirtual class that gets (soft) keyframed to move along with it.

**Reply by wirepair** [reply]
I'd love to see an option for simply adding a rigid body directly to CharacterVirtual. I need sensors to track which characters enter their volume (multiplayer game) for [interest management](https://mirror-networking.gitbook.io/docs/manual/interest-management) purposes. While the latest change can let me see when a character entered a sensor, the sensor doesn't let me know when a character entered it (since it has no BodyID). I am hoping to have these interest management sensors encapsulated in a single interest management class, but without adding a capsule or pairing the CharacterVirtual with a Character I'd need to have some coupling between the character and the interest management class.

**Reply by wirepair** [reply]
If anyone is interested, I wrote a simple guide on how I paired CharcterVirtual with a Capsule rigid body: https://wirepair.org/2024/06/19/area-of-interest-part-2/

**Reply by jrouwe** [reply]
Since f0f19df1ca2f38a8f8f3601859322343b25cc4a2, `CharacterVirtual` can collide with other `CharacterVirtual` through a new interface called `CharacterVsCharacterCollision`. I think this means that it should no longer be needed to have a `Character` attached to a `CharacterVirtual`. `CharacterContactListener` will report both contacts with sensors and contacts with other `CharacterVirtual` now.

**Reply by jrouwe** [reply]
Another change was submitted (d7e8f922d0aa0d073782e9e71ae42b89258c8952) that makes it easier to create an inner rigid body.
